lib/tarball.js

```diff
@@ -0,0 +1,58 @@
+const { relative } = require('path')
+
+const pkgContents = require('@npmcli/installed-package-contents')
+const pacote = require('pacote')
+const isPackageBin = require('pacote/lib/util/is-package-bin.js')
+const tar = require('tar')
+
+// applies same opts as pacote.tarball for directory npa specs
+const tarOpts = manifest => ({
+  cwd: manifest._resolved,
+  prefix: 'package/',
+  portable: true,
+  gzip: true,
+
+  // ensure that package bins are always executable
+  // Note that npm-packlist is already filtering out
+  // anything that is not a regular file, ignored by
+  // .npmignore or package.json "files", etc.
+  filter: (path, stat) => {
+    if (isPackageBin(manifest, path))
+      stat.mode |= 0o111
+    return true
+  },
+
+  // Provide a specific date in the 1980s for the benefit of zip,
+  // which is confounded by files dated at the Unix epoch 0.
+  mtime: new Date('1985-10-26T08:15:00.000Z'),
+})
+
+// returns a simplified tarball when reading files from node_modules folder,
+// thus avoiding running the prepare scripts and the extra logic from packlist
+const nodeModulesTarball = (manifest, opts) =>
+  pkgContents({ path: manifest._resolved, depth: 1 })
+    .then(files =>
+      files.map(file => relative(manifest._resolved, file))
+    )
+    .then(files =>
+      tar.c(tarOpts(manifest), files)
+        .on('error', /* istanbul ignore next */ err => {
+          throw err
+        })
+    )
+    .then(stream => stream.concat())
+
+const tarball = (manifest, opts) => {
+  const resolved = manifest._resolved
+  const where = opts.where || process.cwd()
+
+  const fromNodeModules = resolved.startsWith(where)
+    && relative(where, resolved).startsWith('node_modules')
```

Comment:
What about `./packages/some-workspace/node_modules/a` or `../other-project/node_modules/a`?  It might be safer to just test the path against `/node_modules[\\/](@[^\\/]+\/)?[^\\/]+[\\/]?$/`

Reply:
Hm, looks from CI like this check is failing on Windows as well for some reason, but that's surprising to me.  Not sure why it would be.